winml/adapter/WinMLAdapter.h

```diff
@@ -146,4 +146,60 @@ class InferenceSession : public Microsoft::WRL::RuntimeClass <
     std::shared_ptr<onnxruntime::InferenceSession> session_;
 };
 
-}  // namespace Windows::AI::MachineLearning::Adapter
\ No newline at end of file
+}  // namespace Windows::AI::MachineLearning::Adapter
+
+namespace Ort {
+// Ort::Allocator is not in the C ABI yet so it will have to be in the WinMLAdapter for now.
+// This struct was copied using the Base struct from onnxruntime_cxx_api.h for reference
+// Ort::Allocator struct is used as a smart pointer to OrtAllocator.
+struct Allocator {
+  Allocator() {
+    m_ort_allocator = nullptr;
+    m_adapter = nullptr;
+  }
+  Allocator(winmla::IWinMLAdapter* adapter, OrtAllocator* ort_allocator) :
+      m_adapter(adapter), m_ort_allocator(ort_allocator) {}
+
+  ~Allocator() {
+    if (m_adapter != nullptr && m_ort_allocator != nullptr) {
+      m_adapter->FreeProviderAllocator(m_ort_allocator);
+    }
+  }
+
+  operator OrtAllocator*() { return m_ort_allocator; }
+  operator const OrtAllocator*() const { return m_ort_allocator; }
+
+  OrtAllocator* release() {
+    OrtAllocator* p = m_ort_allocator;
+    m_ort_allocator = nullptr;
+    m_adapter = nullptr;
+    return p;
+  }
+
+  OrtAllocator** put() noexcept {
+    assert(m_ort_allocator == nullptr);
+    return &m_ort_allocator;
+  }
+
+  Allocator(const Allocator&) = delete;
+  Allocator& operator=(const Allocator&) = delete;
+  Allocator(Allocator&& v) noexcept :
+      m_adapter{v.m_adapter}, m_ort_allocator{v.m_ort_allocator} {
+    v.m_adapter = nullptr;
+    v.m_ort_allocator = nullptr;
+  }
+  void operator=(Allocator&& v) noexcept {
+    if (m_ort_allocator != nullptr && m_adapter != nullptr) {
+      m_adapter->FreeProviderAllocator(m_ort_allocator);
+    }
+    m_adapter = v.m_adapter;
+    m_ort_allocator = v.m_ort_allocator;
+    v.m_adapter = nullptr;
+    v.m_ort_allocator = nullptr;
```

Comment:
I think these 4 lines can be replaced with `this = std::move(v)` (or just `this = v`, since it's an rvalue already?) to avoid duplicating the move code.

Reply:
I don't think it's good idea to use std code in header files shared between two DLLs. Let me try and see if I can avoid duplicating the code.

---
In reply to: [358555221](https://github.com/microsoft/onnxruntime/pull/2653#discussion_r358555221) [](ancestors = 358555221)

Reply:
Ok, I think `this = v` works and invokes the move assignment you defined. `std::move` shouldn't pose problems (it's just a cast and shouldn't matter for ABI compatibility) but it's really better to avoid STL altogether